apps/gradient_autoscheduler/GradientAutoscheduler.cpp

```diff
@@ -0,0 +1,1008 @@
+        int max_domain_size = is_gpu ? gpu_max_domain_size : cpu_max_domain_size;
+        if (domain_size < max_domain_size) {
+            if (rvars.size() > 0) {
+                // Check associativity
+                std::vector<Expr> values =
+                    func.update_values(update_id).as_vector();
+                const auto &prover_result =
+                    prove_associativity(func.name(),
+                                        func.update_args(update_id),
+                                        values);
+                // Cache the associative check for later use.
+                checked_associative = true;
+                is_associative = prover_result.associative();
+                if (is_associative) {
+                    schedule_source << func.name() << ".update(" << update_id << ")\n";
+                    // Generate a list of tiled RVars
+                    std::vector<RVar> outer_rvars, inner_rvars;
+                    std::vector<int> outer_rvar_sizes, inner_rvar_sizes;
+                    for (int i = 0; i < (int)rvars.size(); i++) {
+                        if (rvar_bounds[i] >= 8) {
+                            // Let split_size = 8 * n where n is an integer and
+                            // split_size > sqrt(rvar_bounds)
+                            float target = std::sqrt(rvar_bounds[i]);
+                            int split_size = int(std::ceil(target / 8.f)) * 8;
+                            // Split the rvar
+                            RVar outer, inner;
+                            func.update(update_id)
+                                .split(rvars[i], outer, inner, split_size,
+                                       TailStrategy::GuardWithIf);
+                            schedule_source << "    .split(" <<
+                                rvars[i].name() << "," <<
+                                outer.name() << "," <<
+                                inner.name() << "," <<
+                                split_size << "," <<
+                                TailStrategy::GuardWithIf << ")\n";
+                            outer_rvars.push_back(outer);
+                            inner_rvars.push_back(inner);
+                            int outer_size = rvar_bounds[i] % split_size == 0 ? split_size : split_size + 1;
+                            outer_rvar_sizes.push_back(outer_size);
+                            inner_rvar_sizes.push_back(split_size);
+                        } else {
+                            inner_rvars.push_back(rvars[i]);
+                            inner_rvar_sizes.push_back(rvar_bounds[i]);
+                        }
+                    }
+                    schedule_source << ";\n";
+                    if (outer_rvars.size() > 0 && inner_rvars.size() > 0) {
+                        // Rfactor all the outer RVars.
+                        std::vector<std::pair<RVar, Var>> preserved;
+                        std::vector<Var> interm_vars;
```

Comment:
Is this supposed to be "intern_vars" or maybe "interim_vars"?

Reply:
interim, sorry ; )